From ed8931c6e2a57ea6d64014b1e65f896fddd81bc1 Mon Sep 17 00:00:00 2001
From: Vojtech Myslivec <vojtech.myslivec@nic.cz>
Date: Mon, 18 May 2020 19:10:23 +0200
Subject: [PATCH] ppp: Add UCI option to control length of PPPD setup sleep

This patch improves previous 'ppp: add sleep for unreliable DSL (PPP)
connection' patch by introduce `pppd_sleep` UCI option.

`pppd_sleep` option could be used to control the length of "sleeping"
before ppp daemon is started and if set to 0 or less, no sleep would
happened at all. Default value is 10 which match the original hardcoded
value.

This patch is a hack vol.2 and it should not be sent to upstream!

More details are included in:
https://gitlab.labs.nic.cz/turris/turris-build/issues/107
---
 package/network/services/ppp/files/ppp.sh | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/package/network/services/ppp/files/ppp.sh b/package/network/services/ppp/files/ppp.sh
index e44735b9ad..f0831e609d 100755
--- a/package/network/services/ppp/files/ppp.sh
+++ b/package/network/services/ppp/files/ppp.sh
@@ -71,6 +71,7 @@ ppp_generic_init_config() {
 	proto_config_add_string keepalive
 	proto_config_add_boolean keepalive_adaptive
 	proto_config_add_int demand
+	proto_config_add_int pppd_sleep
 	proto_config_add_string pppd_options
 	proto_config_add_string 'connect:file'
 	proto_config_add_string 'disconnect:file'
@@ -92,7 +93,7 @@ ppp_generic_setup() {
 	local config="$1"; shift
 	local localip
 
-	json_get_vars ip6table demand keepalive keepalive_adaptive username password pppd_options pppname unnumbered reqprefix persist maxfail holdoff peerdns sourcefilter delegate norelease
+	json_get_vars ip6table demand keepalive keepalive_adaptive username password pppd_options pppd_sleep pppname unnumbered reqprefix persist maxfail holdoff peerdns sourcefilter delegate norelease
 
 	[ ! -e /proc/sys/net/ipv6 ] && ipv6=0 || json_get_var ipv6 ipv6
 
@@ -145,7 +146,13 @@ ppp_generic_setup() {
 	[ "$sourcefilter" = "0" ] || sourcefilter=""
 	[ "$delegate" != "0" ] && delegate=""
 	[ "$norelease" = "1" ] || norelease=""
-	sleep 10
+
+	# Workaround for unreliable PPPD connection setup
+	[ -n "$pppd_sleep" ] || pppd_sleep=10
+	[ "$pppd_sleep" -gt 0 ] && {
+		echo "ppp: warning: Sleeping for '$pppd_sleep' seconds" >&2
+		sleep "$pppd_sleep"
+	}
 
 	proto_run_command "$config" /usr/sbin/pppd \
 		nodetach ipparam "$config" \
-- 
2.52.0

